src/nimib/sources.nim

```diff
@@ -0,0 +1,202 @@
+
+func getCodeBlock*(source: string, command: string, startPos, endPos: Pos): string =
+  ## Called by getCodeAsInSource
+  let lines = source.split("\n")
+  debugecho "Start line: ", startPos.line
+  var startLine = startPos.line - 1
+  var endLine = endPos.line - 1
+
+  var codeText: string
+  if not lines[startLine].isCommandLine(command): # multiline case
+    while 0 < startLine and not lines[startLine-1].isCommandLine(command):
+      #[ cases like this reports the third line instead of the second line:
+        nbCode:
+          let # this is the line we want
+            x = 1 # but this is the one we get
+      ]#
+      dec startLine
+
+    let indent = skipWhile(lines[startLine], {' '})
+    let indentStr = " ".repeat(indent)
+
+    if lines[endLine].count("\"\"\"") == 1: # only opening of triple quoted string found. Rest is below it. 
+      inc endLine # bump it to not trigger the loop to immediately break
+      while endLine < lines.high and "\"\"\"" notin lines[endLine]:
+        inc endLine
+        debugecho "Triple quote: ", lines[endLine]
+
+    while endLine < lines.high and (lines[endLine+1].startsWith(indentStr) or lines[endLine+1].isEmptyOrWhitespace):# and lines[endLine+1].strip().startsWith("#"):
+      # Ending Comments should be included as well, but they won't be included in the AST -> endLine doesn't take them into account.
+      # Block comments must be properly indented (including the content)
+      inc endLine
+
+    var codeLines = lines[startLine .. endLine]
+
+    var notIndentLines: seq[int] # these lines are not to be adjusted for indentation. Eg content of triple quoted strings.
+    var i: int
+    while i < codeLines.len:
+      if codeLines[i].count("\"\"\"") == 1:
+        # We must do the identification of triple quoted string separatly from the endLine bumping because the triple strings
+        # might not be the last expression in the code block.
+        inc i # bump it to not trigger the loop to immediately break on the initial """
+        notIndentLines.add i
+        while i < codeLines.len and "\"\"\"" notin codeLines[i]:
+          inc i
+          notIndentLines.add i
+      inc i
+      
+
+
+    let parsedLines = collect(newSeqOfCap(codeLines.len)):
+      for i in 0 .. codeLines.high:
+        if i in notIndentLines:
+          codeLines[i]
+        else:
+          codeLines[i].substr(indent)
+    codeText = parsedLines.join("\n")
+    #codeText = codeLines.mapIt(it.substr(indent)).join("\n")
+
+  else: # single line case, eg `nbCode: echo "Hello World"`
+    let line = lines[startLine]
+    codeText = line.split(":")[1 .. ^1].join(":").strip() # split at first ":" and take the rest as code and then strip it.
+  return codeText
+
+macro getCodeAsInSource*(source: static string, command: static string, body: untyped): static string =
+  ## substitute for `toStr` in blocks.nim
+  let startPos = startPos(body)
+  let endPos = finishPos(body)
+  let lObj = body.lineInfoObj()
```

Comment:
this lObj can probably be removed now

Reply:
Yes, was only needed for debugging purposes